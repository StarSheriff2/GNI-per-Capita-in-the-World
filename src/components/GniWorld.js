import React, { useEffect } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { Spinner } from 'react-bootstrap';
import { fetchGniWorld } from '../redux/gniWorld/gniWorld';
import RegionsList from './RegionsList';
import IncomeGroupsList from './IncomesGroupList';

const GniWorld = () => {
  const dispatch = useDispatch();

  const gniWorld = useSelector((state) => state.gniWorld.entities, shallowEqual);
  const loadingStatus = useSelector((state) => state.gniWorld.status);

  useEffect(() => {
    if (gniWorld[0].length === 0) {
      dispatch(fetchGniWorld());
    }
  }, []);

  if (loadingStatus === 'starting') {
    return (
      <Spinner animation="border" role="status">
        <span className="visually-hidden">Loading...</span>
      </Spinner>
    );
  }

  return (
    <div className="gni-world-container">
      <RegionsList regions={gniWorld[0]} />
      <IncomeGroupsList incomeGroups={gniWorld[1]} />
    </div>
  );
};

export default GniWorld;
