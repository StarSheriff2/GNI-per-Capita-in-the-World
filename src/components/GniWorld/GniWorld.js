import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { fetchGniWorld } from '../../redux/gniWorld/gniWorld';
import GroupsList from '../GroupsList/GroupsList';
import Filter from '../Filter/Filter';
import LoadAnimation from '../LoadAnimation/LoadAnimation';
import styles from './GniWorld.module.scss';

const GniWorld = (props) => {
  const dispatch = useDispatch();

  const { currentCategory, updatePath } = props;
  const { current, other } = currentCategory;

  const gniWorld = useSelector((state) => state.gniWorld.entities, shallowEqual);
  const loadingStatus = useSelector((state) => state.gniWorld.status);

  const [categoryFilter, setCategoryFilter] = useState({
    current,
    other,
  });

  const changeCategoryFilter = () => setCategoryFilter((actualCategory) => ({
    current: actualCategory.other,
    other: actualCategory.current,
  }));

  useEffect(() => {
    if (gniWorld.length === 0) {
      dispatch(fetchGniWorld());
    }
  }, []);

  if (loadingStatus === 'starting') {
    return (
      <section className={`${styles.spinnerContainer}`}>
        <LoadAnimation />
      </section>
    );
  }

  return (
    <section>
      <Filter
        currentCategory={categoryFilter.current}
        otherCategory={categoryFilter.other}
        changeCategoryFilter={changeCategoryFilter}
      />
      <div className={`${styles.groupListContainer}`}>
        <GroupsList groups={gniWorld} category={categoryFilter} updatePath={updatePath} />
      </div>
    </section>
  );
};

GniWorld.propTypes = {
  updatePath: PropTypes.func.isRequired,
  currentCategory: PropTypes.shape({
    current: PropTypes.string.isRequired,
    other: PropTypes.string.isRequired,
  }).isRequired,
};

export default GniWorld;
