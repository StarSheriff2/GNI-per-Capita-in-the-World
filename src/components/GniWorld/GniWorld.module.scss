@import '../../styles/abstracts/mixins';

.spinnerContainer {
  @include d-flex(row, center, center);

  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.groupListContainer {
  @include d-flex(row, flex-start, center, wrap);
}
