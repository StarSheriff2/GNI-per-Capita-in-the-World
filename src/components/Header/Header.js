import PropTypes from 'prop-types';
import React from 'react';
import { NavLink } from 'react-router-dom';
import styles from './Header.module.scss';

const Header = (props) => {
  const { currentPath, updatePath } = props;
  const { path } = currentPath;

  let backArrow = (
    <NavLink
      className={styles.backArrow}
      to="/"
      onClick={() => {
        updatePath({
          ...currentPath,
          path: '/',
        });
      }}
    >
      &#60;
    </NavLink>
  );

  let navigationPath = (
    <p className={`${styles.navigationPath}`}>
      {path
        .replace(/groups\//, '')
        .replace(/\//g, '')
        .replace(/-/g, ' ')
        .replace(/and/g, '&')}
    </p>
  );

  if (path === '/') {
    backArrow = (
      <div className="d-none" />
    );
    navigationPath = (
      <p className={`${styles.homePath}`}>GNI per Capita in the World</p>
    );
  }

  return (
    <header id="header-id" className={`d-flex ${styles.header}`}>
      {backArrow}
      {navigationPath}
    </header>
  );
};

Header.propTypes = {
  currentPath: PropTypes.shape({
    path: PropTypes.string.isRequired,
    groupId: PropTypes.string.isRequired,
    currentCategory: PropTypes.shape({
      current: PropTypes.string.isRequired,
      other: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
  updatePath: PropTypes.func.isRequired,
};

export default Header;
