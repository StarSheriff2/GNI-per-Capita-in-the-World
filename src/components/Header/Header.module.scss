@import '../../styles/abstracts/fonts';

.header {
  position: relative;
  padding: 1rem;
}

.backArrow {
  margin-left: 1rem;
  position: absolute;
  left: 0;
}

.navigationPath {
  text-transform: capitalize;
  font-family: $secondary-font;
}

.homePath {
  font-family: $secondary-font;
}
