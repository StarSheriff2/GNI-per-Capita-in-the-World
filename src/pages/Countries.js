// import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { fetchGniWorld } from '../redux/gniWorld/gniWorld';
import { fetchCountries } from '../redux/countries/countries';
import LoadAnimation from '../components/LoadAnimation/LoadAnimation';
import Country from '../components/Country/Country';
import GroupHeader from '../components/GroupHeader/GroupHeader';
import styles from './Countries.module.scss';
import incomeLevelCodes from '../filters/incomeLevel';

const Countries = (props) => {
  const { currentPath, groupId } = props;
  const { currentCategory, path } = currentPath;

  const dispatch = useDispatch();

  const gniWorld = useSelector((state) => state.gniWorld.entities, shallowEqual);
  const groupCountries = useSelector((state) => state.countries.entities, shallowEqual);
  const gniWorldLoadingStatus = useSelector((state) => state.gniWorld.status);
  const countriesLoadingStatus = useSelector((state) => state.countries.status);

  let currentCountries = { fetched: false };

  const validateGroupId = () => {
    if (currentCategory.current === 'region') {
      return groupId;
    }
    return incomeLevelCodes[groupId];
  };

  if (validateGroupId() in groupCountries) {
    const id = validateGroupId();

    currentCountries = {
      fetched: true,
      countries: groupCountries[id].map((country) => {
        let indicator = { value: null, date: null };

        for (let i = 0; i < gniWorld.length; i += 1) {
          if (gniWorld[i].countryiso3code === country.id) {
            indicator = { value: gniWorld[i].value, date: gniWorld[i].date };
            break;
          }
        }

        return ({
          ...country,
          ...indicator,
        });
      }),
    };
  }

  useEffect(() => {
    if (gniWorld.length === 0) {
      dispatch(fetchGniWorld());
    }
    if (gniWorld.length > 0 && !(validateGroupId() in groupCountries)) {
      dispatch(fetchCountries(groupId, currentCategory.current));
    }
  }, [gniWorld]);

  if (gniWorldLoadingStatus === 'starting' || countriesLoadingStatus === 'starting') {
    return (
      <section className={`${styles.spinnerContainer}`}>
        <LoadAnimation />
      </section>
    );
  }

  return (
    <section>
      <GroupHeader
        currentPath={path}
        groupId={groupId}
      />
      <div className="details-container">
        {currentCountries.fetched && currentCountries.countries
          .sort((a, b) => a.value < b.value)
          .map((country) => (
            <li key={country.id} className={`${styles.countryWrapper}`}>
              <Country country={country} />
            </li>
          ))}
      </div>
    </section>
  );
};

/* Details.propTypes = {
  groupId: PropTypes.string.isRequired,
  currentPath: PropTypes.string.isRequired,
}; */

export default Countries;
