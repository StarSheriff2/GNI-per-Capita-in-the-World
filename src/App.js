import {
  HashRouter as Router,
  Switch,
  Route,
} from 'react-router-dom';
import GniWorld from './components/GniWorld';

// import Header from './components/Header';

function App() {
  return (
    <>
      <Router>
        <div className="App">
          <Switch>
            <Route exact path="/">
              <GniWorld />
            </Route>
            {/* <Route path="/Missions">
              <Missions />
            </Route> */}
          </Switch>
        </div>
      </Router>
    </>
  );
}

export default App;
