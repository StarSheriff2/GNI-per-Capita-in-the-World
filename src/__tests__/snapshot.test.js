import React from 'react';
import TestRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import App from '../App';
import store from '../redux/configureStore';
import Header from '../components/Header/Header';
import GniWorld from '../components/GniWorld/GniWorld';
import Countries from '../pages/Countries';

describe('App', () => {
  it('Renders the Homepage', () => {
    const app = TestRenderer.create(
      <Provider store={store}>
        <App />
      </Provider>,
    ).toJSON();
    expect(app).toMatchSnapshot();
  });
});

describe('Header', () => {
  it('Renders the Header', () => {
    const initialPath = {
      path: '/',
      groupId: '',
      currentCategory: {
        current: 'region',
        other: 'income',
      },
    };

    const header = TestRenderer.create(
      <Header currentPath={initialPath} />,
    ).toJSON();
    expect(header).toMatchSnapshot();
  });
});

describe('GniWorld', () => {
  it('Renders the homepage GniWorld', () => {
    const pathX = '/';

    const setPath = jest.fn();

    const func = (newPath) => {
      if (pathX !== newPath) {
        setPath(() => newPath);
      }
    };

    const current = {
      currentCategory: {
        current: 'region',
        other: 'income',
      },
    };

    const gniWorld = TestRenderer.create(
      <Provider store={store}>
        <GniWorld
          updatePath={func}
          currentCategory={current}
        />
      </Provider>,
    ).toJSON();
    expect(gniWorld).toMatchSnapshot();
  });
});

describe('Countries', () => {
  it('Renders the Countries page', () => {
    const groupId = 'EAS';
    const pathX = {
      currentCategory: {
        current: 'region',
        other: 'income',
      },
    };

    const countries = TestRenderer.create(
      <Provider store={store}>
        <Countries
          groupId={groupId}
          currentPath={pathX}
        />
      </Provider>,
    ).toJSON();
    expect(countries).toMatchSnapshot();
  });
});
