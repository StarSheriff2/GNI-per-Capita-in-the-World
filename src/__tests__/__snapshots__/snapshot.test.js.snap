// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Renders the Homepage 1`] = `
<div
  className="App"
>
  <header
    className="d-flex header"
    data-testid="header"
  >
    <div
      className="d-none"
    />
    <p
      className="homePath"
      data-testid="headerText"
    >
      gni per capita in the world
    </p>
  </header>
  <section>
    <div
      className="filterContainer"
    >
      <div
        className="d-flex iconContainer"
      >
        <i
          className="fas fa-globe worldIcon"
        />
      </div>
      <div
        className="filterController"
      >
        <h2
          className="currentCategory"
          data-testid="currentCategory"
        >
          REGIONS
        </h2>
        <p
          className="filterLabel"
        >
          Filter Category: 
        </p>
        <button
          className="filterButton"
          onClick={[Function]}
          type="button"
        >
          INCOME
        </button>
      </div>
    </div>
    <p
      className="indicatorDescription"
    >
      indicators by country region
    </p>
    <div
      className="groupListContainer"
    />
  </section>
</div>
`;

exports[`Details Renders the Details page 1`] = `
<section>
  <div
    className="filterContainer"
  >
    <div
      className="d-flex iconContainer"
    >
      <div
        className="incomeLevelIconDiv"
      >
        <i
          className="fas fa-dollar-sign incomeLevelIcon"
        />
        <i
          className="fas fa-dollar-sign incomeLevelIcon"
        />
        <i
          className="fas fa-dollar-sign incomeLevelIcon"
        />
      </div>
    </div>
    <div
      className="filterController"
    >
      <h2
        className="currentGroup"
      >
        
      </h2>
      <p
        className="groupIndicator"
      >
        $ 0
      </p>
    </div>
  </div>
  <p
    className="indicatorDescription"
  >
    country breakdown
  </p>
  <div
    className="details-container"
  >
    
  </div>
</section>
`;

exports[`GniWorld Renders the homepage GniWorld 1`] = `
<section
  className="spinnerContainer"
>
  <div
    className="spinner-border"
    role="status"
    style={
      Object {
        "height": "14rem",
        "width": "14rem",
      }
    }
  >
    <span
      className="sr-only"
      data-testid="spinnerSpan"
    >
      Loading...
    </span>
  </div>
</section>
`;

exports[`Header Renders the Header 1`] = `
<header
  className="d-flex header"
  data-testid="header"
>
  <div
    className="d-none"
  />
  <p
    className="homePath"
    data-testid="headerText"
  >
    gni per capita in the world
  </p>
</header>
`;
